File: mirror/tier2-python/showcase/pattern_binding.py
"""
MIRROR V3: Tier 2 Python Showcase
Category: Pattern Matching
Module: pattern_binding.py
Purpose: Demonstrate pattern binding and destructuring
"""


def destructure_tuple(data):
    """Destructure tuple into components"""
    if isinstance(data, tuple) and len(data) == 3:
        a, b, c = data
        return {"first": a, "second": b, "third": c}
    return None


def destructure_dict(data):
    """Destructure dict pattern"""
    if isinstance(data, dict):
        name = data.get("name", "unknown")
        age = data.get("age", 0)
        return {"person": name, "age": age}
    return None


def destructure_list(data):
    """Destructure list pattern"""
    if isinstance(data, list) and len(data) >= 2:
        head = data[0]
        tail = data[1:]
        return {"head": head, "tail": tail}
    return None


def bind_and_match(obj):
    """Bind matched values"""
    if isinstance(obj, dict):
        x = obj.get("x")
        y = obj.get("y")
        if x is not None and y is not None:
            return {"bound_x": x, "bound_y": y, "sum": x + y}
    return None


result = {
    "tuple_destructure": destructure_tuple((1, 2, 3)),
    "dict_destructure": destructure_dict({"name": "Alice", "age": 30}),
    "list_destructure": destructure_list([10, 20, 30, 40]),
    "bind_and_match": bind_and_match({"x": 5, "y": 10}),
}

if __name__ == "__main__":
    print(result)

File: mirror/tier2-php/showcase/pattern_binding.php
<?php
// MIRROR V3: Tier 2 PHP Showcase
// Category: Pattern Matching
// Module: pattern_binding.php

function destructure_tuple($data) {
    if (is_array($data) && count($data) === 3) {
        [$a, $b, $c] = $data;
        return ['first' => $a, 'second' => $b, 'third' => $c];
    }
    return null;
}

function destructure_hash($data) {
    if (is_array($data)) {
        $name = $data['name'] ?? 'unknown';
        $age = $data['age'] ?? 0;
        return ['person' => $name, 'age' => $age];
    }
    return null;
}

function destructure_list($data) {
    if (is_array($data) && count($data) >= 2) {
        $head = $data[0];
        $tail = array_slice($data, 1);
        return ['head' => $head, 'tail' => $tail];
    }
    return null;
}

function bind_and_match($obj) {
    if (is_array($obj) && isset($obj['x'], $obj['y'])) {
        $x = $obj['x'];
        $y = $obj['y'];
        return ['bound_x' => $x, 'bound_y' => $y, 'sum' => $x + $y];
    }
    return null;
}

$result = [
    'tuple_destructure' => destructure_tuple([1, 2, 3]),
    'hash_destructure' => destructure_hash(['name' => 'Alice', 'age' => 30]),
    'list_destructure' => destructure_list([10, 20, 30, 40]),
    'bind_and_match' => bind_and_match(['x' => 5, 'y' => 10]),
];

echo json_encode($result);
?>
